src/rules/dollar-variable-pattern/index.js

```diff
@@ -0,0 +1,35 @@
+import { isRegExp, isString } from "lodash"
+import { utils } from "stylelint"
+
+export const ruleName = "dollar-variable-pattern"
+
+export const messages = utils.ruleMessages(ruleName, {
+  expected: "Expected $ variable name to match specified pattern",
+})
+
+export default function (pattern) {
+  return (root, result) => {
+    const validOptions = utils.validateOptions(result, ruleName, {
+      actual: pattern,
+      possible: [ isRegExp, isString ],
+    })
+    if (!validOptions) { return }
+
+    const regexpPattern = (isString(pattern))
+      ? new RegExp(pattern)
+      : pattern
+
+    root.walkDecls(decl => {
+      const { prop } = decl
+      if (prop.slice(0, 1) !== "$") { return }
```

Comment:
True. Just tested `slice` to make sure, it is about 20% slower, but indeed for 100000 operations it makes only about 20ms in modern browsers. Yet, not the best code, sorry, my bad